Build failure: python3.11-azure-cli

### Steps To Reproduce

Steps to reproduce the behavior:
1. build azure-cli

### Build log

```
error: builder for '/nix/store/qbk4jfwi7gyhfy5qdynwl52fsgh3flnm-python3.11-azure-cli-core-2.53.0.drv' failed with exit code 1;
       last 50 log lines:
       > azure/cli/core/tests/test_aaz_operation.py:402
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_aaz_operation.py:402: DeprecationWarning: invalid escape sequence '\['
       >     with self.assertRaisesRegex(InvalidArgumentValueError, "Couldn't find 'o' in 'l\[0\]'. Available options: \['p', 's'\]"):
       >
       > azure/cli/core/tests/test_aaz_operation.py:444
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_aaz_operation.py:444: DeprecationWarning: invalid escape sequence '\['
       >     with self.assertRaisesRegex(InvalidArgumentValueError, "Couldn't find 't' in 'l\[1\]'. Available options: \['p', 's'\]"):
       >
       > azure/cli/core/tests/test_aaz_operation.py:464
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_aaz_operation.py:464: DeprecationWarning: invalid escape sequence '\('
       >     with self.assertRaisesRegex(InvalidArgumentValueError, "Expect <class 'str'>, got 3 \(<class 'int'>\)"):
       >
       > azure/cli/core/tests/test_aaz_operation.py:484
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_aaz_operation.py:484: DeprecationWarning: invalid escape sequence '\['
       >     with self.assertRaisesRegex(InvalidArgumentValueError, "Couldn't find '\[10\]' in 'd'. Available options: \['a', 'b', 'c'\]"):
       >
       > azure/cli/core/tests/test_aaz_operation.py:495
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_aaz_operation.py:495: DeprecationWarning: invalid escape sequence '\('
       >     with self.assertRaisesRegex(InvalidArgumentValueError, "Expect <class 'int'>, got a \(<class 'str'>\)"):
       >
       > azure/cli/core/extension/__init__.py:11
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/extension/__init__.py:11: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
       >     from distutils.sysconfig import get_python_lib  # pylint: disable=deprecated-module
       >
       > azure/cli/core/extension/__init__.py:11
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/extension/__init__.py:11: DeprecationWarning: The distutils.sysconfig module is deprecated, use sysconfig instead
       >     from distutils.sysconfig import get_python_lib  # pylint: disable=deprecated-module
       >
       > azure/cli/core/tests/test_help_loaders.py:13
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_help_loaders.py:13: PytestCollectionWarning: cannot collect test class 'TestCommandLoader' because it has a __init__ constructor (from: azure/cli/core/tests/test_help_loaders.py)
       >     class TestCommandLoader(AzCommandsLoader):
       >
       > azure/cli/core/tests/test_application.py::TestApplication::test_client_request_id_is_refreshed_after_execution
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_application.py:56: DeprecationWarning: Please use assertNotEqual instead.
       >     self.assertNotEquals(old_id, cli.data['headers']['x-ms-client-request-id'])
       >
       > azure/cli/core/tests/test_application.py::TestApplication::test_client_request_id_is_refreshed_correctly
       >   /build/azure-cli-2.53.0-src/src/azure-cli-core/azure/cli/core/tests/test_application.py:35: DeprecationWarning: Please use assertNotEqual instead.
       >     self.assertNotEquals(old_id, cli.data['headers']['x-ms-client-request-id'])
       >
       > azure/cli/core/tests/test_command_registration.py: 10 warnings
       >   /nix/store/ffll6glz3gwx342z0ch8wx30p5cnqz1z-python3-3.11.5/lib/python3.11/unittest/case.py:1176: DeprecationWarning: assertDictContainsSubset is deprecated
       >     warnings.warn('assertDictContainsSubset is deprecated',
       >
       > -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
       > =========================== short test summary info ============================
       > FAILED azure/cli/core/tests/test_command_registration.py::TestCommandRegistration::test_command_index - AttributeError: <MagicMock id='140737265805712'> does not have the attribut...
       > FAILED azure/cli/core/tests/test_command_registration.py::TestCommandRegistration::test_command_index_always_loaded_extensions - AssertionError: [] != 'FAKE_HANDLER'
       > FAILED azure/cli/core/tests/test_parser.py::TestParser::test_parser_error_spellchecker - AssertionError: 0 != 5
       > =========== 3 failed, 237 passed, 3 deselected, 22 warnings in 6.15s ===========
       For full logs, run 'nix log /nix/store/qbk4jfwi7gyhfy5qdynwl52fsgh3flnm-python3.11-azure-cli-core-2.53.0.drv'.
error: 1 dependencies of derivation '/nix/store/2fnf2n69hxkair57l0jm4fjxlsbn3s4a-python3.11-azure-cli-2.53.0.drv' failed to build

```

### Additional context

Add any other context about the problem here.

### Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
🔨 Welcome to khanelinix
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.90.1-microsoft-standard-WSL2, NixOS, 23.11 (Tapir), 23.11.20231019.7c9cc5a`
 - multi-user?: `yes`
 - sandbox: `relaxed`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/etc/nix/inputs/nixpkgs`

```

--- a/pkgs/tools/admin/azure-cli/default.nix
+++ b/pkgs/tools/admin/azure-cli/default.nix
@@ -1,7 +1,5 @@
 { lib
-, stdenv
-, python3
-, fetchPypi
+, callPackage
 , fetchFromGitHub
 , installShellFiles
 }:
@@ -18,9 +16,7 @@ let
   };
 
   # put packages that needs to be overridden in the py package scope
-  py = import ./python-packages.nix {
-    inherit stdenv src version python3 fetchPypi;
-  };
+  py = callPackage ./python-packages.nix { inherit src version; };
 in
 
 py.pkgs.toPythonApplication (py.pkgs.buildAzureCliPackage {

--- a/pkgs/tools/admin/azure-cli/python-packages.nix
+++ b/pkgs/tools/admin/azure-cli/python-packages.nix
@@ -1,6 +1,7 @@
 { stdenv
 , python3
 , fetchPypi
+, fetchpatch
 , src
 , version
 }:
@@ -27,6 +28,20 @@ let
       azure-cli-core = buildAzureCliPackage {
         pname = "azure-cli-core";
         inherit version src;
+
+        patches = [
+          (fetchpatch {
+            name = "fix-python311.patch";
+            url = "https://github.com/Azure/azure-cli/commit/a5198b578b17de934e15b1c92e369e45323e9658.patch";
+            hash = "sha256-qbyKF6Vvtz8QwY78sG7ptTVcbM2IR+phntOKqsrWetE=";
+            stripLen = 2;
+            includes = [
+              "azure/cli/core/tests/test_command_registration.py"
+              "azure/cli/core/tests/test_help.py"
+              "azure/cli/core/tests/test_parser.py"
+            ];
+          })
+        ];
 
         sourceRoot = "${src.name}/src/azure-cli-core";
 

